jina/types/arrays/memmap.py

```diff
@@ -214,15 +326,27 @@ def __iter__(self) -> Iterator['Document']:
     def __setitem__(self, key: Union[int, str], value: 'Document') -> None:
         if isinstance(key, int):
             if 0 <= key < len(self):
+                str_key = self._int2str_id(key)
                 # override an existing entry
-                self.append(value)
-                self._header_map[self._int2str_id(key)] = self._header_map[value.id]
-                del self[value.id]
+                self.update(value, key)
+
+                # allows overwriting an existing document
+                if str_key != value.id:
+                    entry = self._header_map.pop(value.id)
+                    self._header_map = OrderedDict(
+                        [
+                            (value.id, entry) if k == str_key else (k, v)
+                            for k, v in self._header_map.items()
+                        ]
+                    )
```

Comment:
This is super expensive. is there any reason why this is needed?

Reply:
yes, the problem is when we want to call setitem with another document (different ID), we need to change the key in the header map but keep the same order (because __iter__ relies on the order).
I don't know a datastructure that supports inserting at some position efficiently

Reply:
Can you point me to a test that fails, when we don't use the `OrderedDict`?

Reply:
I don't think there are any. Now I'm aware that using a normal dict is fine.
I can point you to a case where updating at this point without reconstructing the dict corrupts the order of the entries

Reply:
Here's a valid reason why _header_map should be an ordered dict:
due to this issue: https://github.com/jina-ai/jina/issues/3076, when adding a new entry to the header map, the idx should not be the len but rather the largest idx + 1
I think the most efficient way is to use next(reversed(self._header_map.items()))
in python3.7, this is not supported for generic dicts: 
https://stackoverflow.com/questions/58413076/why-are-python-dictionaries-not-reversible-for-python3-7